PYME/recipes/localisations.py

```diff
@@ -147,7 +168,16 @@ def execute(self, namespace):
             mapped_ds.setMapping('y', 'y*pixelSize')
 
         #extract information from any events
-        events = namespace.get(self.inputEvents, None)
+        if self.inputEvents != '':
+            logger.warn('Having Events in the recipe namespace is deprecated, they should be associated with datasources during load, or with PYME.recipes.fudges.AddEvents')
```

Comment:
```suggestion
            # Use specified table for events if given (otherwise look for a `.events` attribute on the input data
            # TODO: resolve how best to handle events (i.e. should they be a separate table, or should they be attached to data tables)
```

I know we discussed this, but don't believe there was a final conclusion. There are a couple of competing considerations here - having events in the namespace results in clutter, but is conversely the most generic way of handling them - particularly if we want to minimise custom file formats. There is also a general ugliness with attaching attributes to tabular data (and deciding if and when to propagate them). 

As I recall, I think I was tending to not really wanting them in the VisGUI namespace (to avoid them appearing as a potential output and the confusion this would generate), but I don't think this automatically extended to recipes and the bakeshop. I definitely don't think that events should be propagated by default in recipe modules (which makes the order of modules somewhat critical - i.e. `Pipelineify` must come before any other modules - a behaviour which is a little counterintuitive). A reasonable compromise is to handle both for now.

Reply:
Once events are loaded and you're ready to inject them into the namespace there is no file format, which i think fundamentally invalidates that concern. They can be loaded however you like, but what do Events mean without an associated datasource, and what happens when you have e.g. localizations and a widefield stack with different events in the same recipe? I think events floating without association are fundamentally as weird as injecting a metadata handler into the recipe namespace. Propagation wise they should be treated similarly too - if the events are invalidated by e.g. averaging frames / changing the length of an image stack, the output imagestack should either have spoofed events or no events.

Reply:
Putting them into the namespace has some advantages re different events in the same recipe and/or events with different file backends (the logical way of handling, e.g. a .tif stack with events or a .dcimg with events would be to put the events in an accompanying .json). If we strictly attach events to datasources on load we would need to follow some pretty prescriptive file naming patterns and have a lot of logic in the loaders to cope with that. If we can just load a json in separately as an events table (different events tables can have different names) then the custom logic for a special case config resides in the recipe that processes it. 

Not saying that this is ideal, but attaching them to data also has it's issues (one of which is how it impacts serialisation). Philosophically I'd prefer it if people didn't need to think about events when writing / using recipe modules - apart from the one or two cases where you need them. This argues against having events be an expected attribute of a tabular object.

Reply:
As you said, we both know we've discussed this, and as mentioned [above in this PR discussion](https://github.com/python-microscopy/python-microscopy/pull/284#discussion_r444630231) I think it would be perfectly reasonable and sane to write an `AddEvents` module to associate events with a datasource. I don't think it's wild to ask a user to shim in their tif/dcimg acquisition events with a specific filename or manually select/indicate it. This is what we did for high-throughput dcimgspooling.